libaums/src/main/java/com/github/mjdev/libaums/driver/scsi/ScsiBlockDevice.kt

```diff
@@ -133,45 +135,74 @@ class ScsiBlockDevice(private val usbCommunication: UsbCommunication, private va
      * The command which should be transferred.
      * @param inBuffer
      * The buffer used for reading or writing.
+     * @param writeCommand
+     * If this flag is true the caller doesn't expect any result in inBuffer
      * @return True if the transaction was successful.
      * @throws IOException
      * If something fails.
      */
     @Throws(IOException::class)
-    private fun transferCommand(command: CommandBlockWrapper, inBuffer: ByteBuffer): Boolean {
+    private fun transferCommand(command: CommandBlockWrapper, inBuffer: ByteBuffer,
+                                writeCommand: Boolean = false) {
         for(i in 0..MAX_RECOVERY_ATTEMPTS) {
             try {
-                return transferOneCommand(command, inBuffer)
-            } catch(e: IOException) {
-                Log.e(TAG, "Error transferring command; errno ${ErrNo.errno} ${ErrNo.errstr}")
-
-                // Try alternately to clear halt and reset device until something happens
-                when {
-                    i == MAX_RECOVERY_ATTEMPTS -> {
-                        Log.d(TAG, "Giving up")
-                        throw e
-                    }
-                    i % 2 == 0 -> {
-                        Log.d(TAG, "Reset bulk-only mass storage")
-                        bulkOnlyMassStorageReset()
-                        Log.d(TAG, "Trying to clear halt on both endpoints")
-                        usbCommunication.clearFeatureHalt(usbCommunication.inEndpoint)
-                        usbCommunication.clearFeatureHalt(usbCommunication.outEndpoint)
-                    }
-                    i % 2 == 1 -> {
-                        Thread.sleep(300 * i.toLong())
-                        Log.d(TAG, "Trying to reset the device")
-                        usbCommunication.resetDevice()
-                    }
-                }
+                val result = transferOneCommand(command, inBuffer)
+                handleCommandResult(result)
+                return
+            } catch(e: Recovered) {
+
+                Log.i(TAG, e.message ?: "Recovered exception caught")
+
+                // If it was a write command the Recovered exception indicates that it
+                // has completed successfully, but if it was a read command we would have to
+                // run it again to receive the correct data.
+                if (writeCommand)
```

Comment:
This writeCommand is only true for the test unit ready command, is this correct?

Is it possible to ditch this flag and use the direction of the command block wrapper like in the rest of the code?